src/function/gds/strongly_connected_components.cpp

```diff
@@ -0,0 +1,245 @@
+        componentIDs = componentIDsMap.getData(tableID);
+        visits = visitsMap.getData(tableID);
+    }
+
+    bool updateVisit(common::offset_t offset, bool value) {
+        visits[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    bool getVisit(offset_t offset) { return visits[offset].load(std::memory_order_relaxed); }
+
+    bool updateComponentID(common::offset_t offset, offset_t value) {
+        componentIDs[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    offset_t getComponentID(offset_t offset) {
+        return componentIDs[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    // Pointers to current tableID's visits and componentIDs
+    std::atomic<bool>* visits = nullptr;
+    std::atomic<offset_t>* componentIDs = nullptr;
+    // State for visits and componentIDs
+    ObjectArraysMap<std::atomic<bool>> visitsMap;
+    ObjectArraysMap<std::atomic<offset_t>> componentIDsMap;
+};
+
+class SCCCompute {
+public:
+    SCCCompute(graph::Graph* graph, const table_id_map_t<offset_t>& numNodesMap, SCCState& sccState)
+        : graph{graph}, sccState{sccState}, numNodesMap{numNodesMap} {};
+    ~SCCCompute() = default;
+
+    // Forward DFS
+    void visit_vertex(nodeID_t node, catalog::TableCatalogEntry* relEntry) {
+        if (sccState.getVisit(node.offset)) {
+            // Already visited
+            return;
+        };
+        sccState.updateVisit(node.offset, true);
+        auto scanState = graph->prepareRelScan(relEntry, "");
+        for (auto chunk : graph->scanFwd(node, *scanState)) {
+            chunk.forEach([&](auto nbrNodeID, auto) { visit_vertex(nbrNodeID, relEntry); });
+        }
+        queue.push_back(node);
+    }
+
+    // Backwards DFS
+    void assign_id(nodeID_t node, offset_t root, catalog::TableCatalogEntry* relEntry) {
+        if (sccState.getComponentID(node.offset) != NOT_ASSIGNED) {
+            // Already assigned
+            return;
+        };
+        sccState.updateComponentID(node.offset, root);
+        auto scanState = graph->prepareRelScan(relEntry, "");
+        for (auto chunk : graph->scanBwd(node, *scanState)) {
+            chunk.forEach([&](auto nbrNodeID, auto) { assign_id(nbrNodeID, root, relEntry); });
+        }
+    }
+
+    void compute() {
+        for (const auto& [tableID, numNodes] : numNodesMap) {
+            sccState.pinTable(tableID);
+            for (auto& [fromEntry, toEntry, relEntry] : graph->getRelFromToEntryInfos()) {
+                for (auto i = 0u; i < numNodes; ++i) {
+                    auto nodeID = nodeID_t{i, tableID};
+                    visit_vertex(nodeID, relEntry);
+                }
+                for (auto it = queue.rbegin(); it != queue.rend(); ++it) {
+                    auto node = *it;
+                    assign_id(node, node.offset, relEntry);
+                }
+            }
+        }
+    }
+
+private:
+    std::vector<nodeID_t> queue;
```

Comment:
I wonder if this will perform if the size of nodes is 1B. Let's see it's performance first.